### PR TITLE
feat: prevent cyclic jointures between pipelines

### PR DESCRIPTION
**Pull request by erenard**
This PR prevent cyclic references between pipelines during the `domain`, `join` and `append` steps.

### CODE REVIEW

**Comment by erenard** [comment]
[(5) ETQC je ne peux faire une jointure que sur un dataset qui ne crée pas de référence cyclique.](https://trello.com/c/7IwwBZyN/609-5-etqc-je-ne-peux-faire-une-jointure-que-sur-un-dataset-qui-ne-cr%C3%A9e-pas-de-r%C3%A9f%C3%A9rence-cyclique)

File: src/store/utils/dereference-pipeline.ts [hunk 1]
```diff
@@ -60,3 +60,112 @@ export function dereferencePipelines(
   }
   return dereferencedPipeline;
 }
+
+function getStepReferences(step: PipelineStep): Reference[] {
```
Comment (alice-sevin) [inline src/store/utils/dereference-pipeline.ts:64]: Cristal clear :heart:

File: src/store/utils/dereference-pipeline.ts [hunk 2]
```diff
@@ -60,3 +60,112 @@ export function dereferencePipelines(
...
+ * @param      {Map}                    cachedResults A map of intermédiate results
+ */
+function _getPipelineNamesReferencedBy(
+  reference: Reference,
+  pipelines: PipelinesScopeContext,
+  cachedResults: Map<string, Array<string>>,
+): Array<string> {
+  const referenceNames: Array<string> = [];
+  if (typeof reference === 'string') {
+    // if reference is a String
```
Comment (alice-sevin) [inline src/store/utils/dereference-pipeline.ts:94]: May no need all the comments as condition are smart but better more than less!

File: src/store/utils/dereference-pipeline.ts [hunk 3]
```diff
@@ -60,3 +60,112 @@ export function dereferencePipelines(
...
+ * Gets the names of all the pipelines referenced by the given one.
+ *
+ * @param      {Reference}              reference  The given pipeline or pipeline name
+ * @param      {PipelinesScopeContext}  pipelines  All the known pipelines
+ */
+export function getPipelineNamesReferencedBy(
+  reference: Reference,
+  pipelines: PipelinesScopeContext,
+): Array<string> {
+  // This method is only wrap the private one to hide the cache from outside this module.
```
Comment (alice-sevin) [inline src/store/utils/dereference-pipeline.ts:169]: :+1:

File: tests/unit/dereference-pipeline.spec.ts [hunk 4]
```diff
@@ -130,3 +135,117 @@ describe('dereferencePipelines', () => {
...
+      it("should add the referenced pipeline's name", () => {
+        expect(
+          getPipelineNamesReferencedBy(
+            [{ name: 'append', pipelines: ['dataset1'] }],
+            pipelinesScopeContext,
+          ),
+        ).toEqual(['dataset1', 'toto']);
+      });
+    });
+    describe('having a none of [domain, append, join] step', () => {
```
Comment (alice-sevin) [inline tests/unit/dereference-pipeline.spec.ts:189]: Clean one!

File: tests/unit/store.spec.ts [hunk 5]
```diff
@@ -9,6 +9,16 @@ import { currentPipeline, emptyState } from '@/store/state';
 import { buildState, buildStateWithOnePipeline, setupMockStore } from './utils';
 
 describe('getter tests', () => {
+  describe('pipelines', () => {
```
Comment (alice-sevin) [inline tests/unit/store.spec.ts:12]: :+1:

File: tests/unit/store.spec.ts [hunk 6]
```diff
@@ -64,6 +74,55 @@ describe('getter tests', () => {
...
+        currentPipelineName: 'coco_l_asticot',
+        pipelines: {
+          coco_l_asticot: [],
+          dataset1: [],
+          dataset2: [],
+        },
+      });
+      expect(getters.availablePipelines(state, {}, {}, {})).toEqual(['dataset1', 'dataset2']);
+    });
+    it('should return everything if the currentPipelineName is not defined', () => {
```
Comment (alice-sevin) [inline tests/unit/store.spec.ts:89]: :+1:

File: tests/unit/multiselect-widget.spec.ts [hunk 7]
```diff
@@ -1,68 +1,4 @@
...
-        value: ['Foo'],
-      },
-    });
-    const inputWrapper = wrapper.find('input[type="text"]');
-    (inputWrapper.element as HTMLInputElement).value = 'Bar';
-    inputWrapper.trigger('input', { value: 'Bar' });
-    expect(wrapper.emitted()).toEqual({ input: [['Bar']] });
-  });
-});
+describe('Widget Multiselect', () => {});
```
Comment (alice-sevin) [inline tests/unit/multiselect-widget.spec.ts]: ??? we may need to keep previous tests in anyway no ?

File: src/components/stepforms/AppendStepForm.vue [hunk 8]
```diff
@@ -5,10 +5,13 @@
...
       v-model="editedStep.pipelines"
       name="Select datasets to append:"
-      :options="[...availablePipelines, ...domains]"
+      :options="options"
       placeholder="Select datasets"
       data-path=".pipelines"
       :errors="errors"
+      track-by="trackBy"
+      label="label"
+      with-example
```
Comment (alice-sevin) [inline src/components/stepforms/AppendStepForm.vue:14]: What does the with-example directive is used for ? Did'nt find it in vue-multiselect documentation ?

File: src/components/stepforms/widgets/Multiselect.vue [hunk 9]
```diff
@@ -5,11 +5,25 @@
...
       :close-on-select="false"
       openDirection="bottom"
-    />
+    >
+      <!-- If you want to use those templates you should provide a 'label' and 
+    'example' key in the options-->
+      <template v-if="withExample" slot="singleLabel" slot-scope="props">
+        <span class="option__title">{{ props.option.label }}</span>
+      </template>
+      <template v-if="withExample" slot="option" slot-scope="props">
```
Comment (alice-sevin) [inline src/components/stepforms/widgets/Multiselect.vue:20]: Alright found answer to one of my previous questions!

**Pull request merged**